### PR TITLE
Added Zawy LWMA difficulty algorithm

### PR DESCRIPTION
**Pull request by wilsonmeier**

### CODE REVIEW

File: lib/bitcoin.py [hunk 1]
```diff
@@ -77,104 +78,142 @@ class NetworkConstants:
...
         cls.GENESIS = "0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206"
+
         cls.DEFAULT_PORTS = {'t': '50001', 's': '50002'}
         cls.DEFAULT_SERVERS = read_json('servers.json', {})
         cls.CHECKPOINTS = read_json('checkpoints_regtest.json', [])
-        cls.FORK_HEIGHT = 3000
+
+        cls.BTG_HEIGHT = 2000
+        cls.ZAWY_HEIGHT = -1
+        #cls.BTG_HEIGHT = 3000
```
Comment (h4x3rotab) [inline lib/bitcoin.py]: Don't leave commented out code in prod code.

File: lib/blockchain.py [hunk 2]
```diff
@@ -419,89 +418,153 @@ def get_hash(self, height):
...
+        if height == 0:
             new_target = NetworkConstants.POW_LIMIT_LEGACY
+        # Check for valid checkpoint
         elif height % difficulty_adjustment_interval() == 0 and 0 <= ((height // difficulty_adjustment_interval()) - 1) < len(self.checkpoints):
             h, t = self.checkpoints[((height // difficulty_adjustment_interval()) - 1)]
             new_target = t
-        elif is_postfork(height):
-            new_target = self.get_postfork_target(height, headers)
+        # Check for prefork
+        elif not is_postfork(height):
```
Comment (h4x3rotab) [inline lib/blockchain.py]: To keep consistent with the code below, better use `height < NetworkConstants.BTG_HEIGHT` instead.

File: lib/bitcoin.py [hunk 3]
```diff
@@ -77,104 +78,142 @@ class NetworkConstants:
...
-        cls.POW_MAX_ADJUST_DOWN = 16
-        cls.POW_MAX_ADJUST_UP = 32
         cls.POW_TARGET_SPACING = 10 * 60
         cls.POW_TARGET_TIMESPAN_LEGACY = 14 * 24 * 60 * 60
 
+        cls.DIGI_AVERAGING_WINDOW = 30
+        cls.DIGI_MAX_ADJUST_DOWN = 32
+        cls.DIGI_MAX_ADJUST_UP = 16
+
+        cls.ZAWY_AVERAGING_WINDOW = 45
```
Comment (h4x3rotab) [inline lib/bitcoin.py]: Better to replace `ZAWY_*` by `LWMA_*`

File: lib/blockchain.py [hunk 4]
```diff
@@ -419,89 +418,153 @@ def get_hash(self, height):
...
-    def get_postfork_target(self, height, headers=None):
-        # Premine
-        if (height < NetworkConstants.FORK_HEIGHT + NetworkConstants.PREMINE_SIZE) or NetworkConstants.REGTEST:
+    def get_zawy_lwma_target(self, height, headers):
+        cur = headers[height] if height in headers else self.read_header(height)
+        last_height = (height - 1)
+        last = headers[last_height] if last_height in headers else self.read_header(last_height)
+
+        # Special testnet handling
+        if (NetworkConstants.TESTNET or NetworkConstants.REGTEST) and cur.get('timestamp') > last.get('timestamp') + NetworkConstants.POW_TARGET_SPACING * 2:
```
Comment (h4x3rotab) [inline lib/blockchain.py]:
```
if regtest:
    return prev_target
elif testnet and itme >= 20min:
    return POW_LIMIT

# the regular logic follows
```

File: lib/blockchain.py [hunk 5]
```diff
@@ -419,89 +418,153 @@ def get_hash(self, height):
...
 
-            pow_limit = NetworkConstants.POW_LIMIT
-            height -= 1
-            last = headers[height] if height in headers else self.read_header(height)
+            # Loop through N most recent blocks.  "< height", not "<=".
+            # height-1 = most recently solved block
+            for i in range(height - NetworkConstants.ZAWY_AVERAGING_WINDOW, height):
+                cur = headers[i] if i in headers else self.read_header(i)
+                prev_height = (i - 1)
+                prev = headers[prev_height] if prev_height in headers else self.read_header(prev_height)
```
Comment (h4x3rotab) [inline lib/blockchain.py]: There are lots of `headers[h] if h in headers else self.read_header(h)`. Maybe we should make it a common function.

File: lib/blockchain.py [hunk 6]
```diff
@@ -419,89 +418,153 @@ def get_hash(self, height):
...
+            total = 0
+            t = 0
+            j = 0
 
-            pow_limit = NetworkConstants.POW_LIMIT
-            height -= 1
-            last = headers[height] if height in headers else self.read_header(height)
+            # Loop through N most recent blocks.  "< height", not "<=".
+            # height-1 = most recently solved block
+            for i in range(height - NetworkConstants.ZAWY_AVERAGING_WINDOW, height):
```
Comment (h4x3rotab) [inline lib/blockchain.py]: Assert `height - NetworkConstants.ZAWY_AVERAGING_WINDOW > 0`

File: lib/blockchain.py [hunk 7]
```diff
@@ -419,89 +418,153 @@ def get_hash(self, height):
...
+        pow_limit = NetworkConstants.POW_LIMIT
+        height -= 1
+        last = headers[height] if height in headers else self.read_header(height)
 
-                if avg > pow_limit:
-                    avg = pow_limit
+        if last is None:
+            new_target = pow_limit
+        elif NetworkConstants.REGTEST:
+            new_target = self.bits_to_target(last.get('bits'))
```
Comment (h4x3rotab) [inline lib/blockchain.py]: We have `self.get_prev_target `?

Comment (wilsonmeier) [reply lib/blockchain.py]: No, was a merge/rebase left over.

**Comment by h4x3rotab** [comment]
lgtm

**Comment by h4x3rotab** [comment]
Have we tested LWMA on testnet before?

**Pull request merged**